lib/membrane_element_aac/encoder.ex

```diff
@@ -0,0 +1,149 @@
+defmodule Membrane.Element.AAC.Encoder do
+  @moduledoc """
+  Element encoding raw audio into AAC format
+  """
+
+  use Membrane.Element.Base.Filter
+  alias __MODULE__.Native
+  alias Membrane.Buffer
+  alias Membrane.Caps.Audio.Raw
+  alias Membrane.Event.EndOfStream
+
+  @channels 2
+  @sample_size 2
+  @default_sample_rate 44_100
+  @aac_frame_size 1024
+  # TODO: 2 is for AAC LC, Add handling different AOTs
+  @audio_object_type 2
+
+  def_output_pads output: [
+                    caps: :any
+                  ]
+
+  def_input_pads input: [
+                   demand_unit: :bytes,
+                   caps:
+                     {Raw, format: :s16le, sample_rate: @default_sample_rate, channels: @channels}
+                 ]
+
+  @impl true
+  def handle_init(options) do
+    {:ok,
+     %{
+       native: nil,
+       options: options,
+       queue: <<>>
+     }}
+  end
+
+  @impl true
+  def handle_stopped_to_prepared(_ctx, state) do
+    with {:ok, native} <-
+           Native.create(
+             @channels,
+             @default_sample_rate,
+             @audio_object_type
+           ) do
+      {:ok, %{state | native: native}}
+    else
+      {:error, reason} -> {{:error, reason}, state}
+    end
+  end
+
+  @impl true
+  def handle_prepared_to_stopped(_ctx, state) do
+    {:ok, %{state | native: nil}}
+  end
+
+  @impl true
+  def handle_demand(:output, size, :bytes, _ctx, state) do
+    {{:ok, demand: {:input, size}}, state}
+  end
+
+  @impl true
+  def handle_demand(:output, bufs, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, @aac_frame_size * bufs}}, state}
+  end
+
+  @impl true
+  def handle_caps(:input, _caps, _ctx, state) do
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_process(:input, %Buffer{payload: data}, _ctx, state) do
+    %{native: native, queue: queue} = state
+
+    to_encode = queue <> data
+
+    with {:ok, {encoded_buffers, bytes_used}} when bytes_used > 0 <-
+           encode_buffer(to_encode, native) do
+      <<_handled::binary-size(bytes_used), rest::binary>> = to_encode
+
+      buffer_actions = [buffer: {:output, encoded_buffers}]
+
+      {{:ok, buffer_actions ++ [redemand: :output]}, %{state | queue: rest}}
+    else
+      {:ok, {[], 0}} -> {:ok, %{state | queue: to_encode}}
+      {:error, reason} -> {{:error, reason}, state}
+    end
+  end
+
+  @impl true
+  def handle_event(:input, %EndOfStream{}, _ctx, state) do
+    %{native: native} = state
+
+    # TODO: Is it possible for state.queue to still contain data here?
```

Comment:
It is, but only if the input is somehow broken, so such situation could issue a warning